packages/node/src/lib/token-manager.ts

```diff
@@ -0,0 +1,328 @@
+import { uuid } from './uuid'
+import {
+  FetchHTTPClient,
+  HTTPClient,
+  HTTPClientRequest,
+  HTTPResponse,
+} from './http-client'
+import { SignOptions, sign } from 'jsonwebtoken'
+import { Emitter, backoff, sleep } from '@segment/analytics-core'
+import type {
+  AccessToken,
+  OAuthSettings,
+  TokenManager as ITokenManager,
+} from './types'
+
+const isAccessToken = (thing: unknown): thing is AccessToken => {
+  return Boolean(
+    thing &&
+      typeof thing === 'object' &&
+      'access_token' in thing &&
+      'expires_in' in thing &&
+      typeof thing.access_token === 'string' &&
+      typeof thing.expires_in === 'number'
+  )
+}
+
+const isValidCustomResponse = (
+  response: HTTPResponse
+): response is HTTPResponse & Required<Pick<HTTPResponse, 'text'>> => {
+  return typeof response.text === 'function'
+}
+
+function convertHeaders(
+  headers: HTTPResponse['headers']
+): Record<string, string> {
+  const lowercaseHeaders: Record<string, string> = {}
+  if (!headers) return {}
+  if (isHeaders(headers)) {
+    for (const [name, value] of headers.entries()) {
+      lowercaseHeaders[name.toLowerCase()] = value
+    }
+    return lowercaseHeaders
+  }
+  for (const [name, value] of Object.entries(headers)) {
+    lowercaseHeaders[name.toLowerCase()] = value as string
+  }
+  return lowercaseHeaders
+}
+
+function isHeaders(thing: unknown): thing is HTTPResponse['headers'] {
+  if (
+    typeof thing === 'object' &&
+    thing !== null &&
+    'entries' in Object(thing) &&
+    typeof Object(thing).entries === 'function'
+  ) {
+    return true
+  }
+  return false
+}
+
+export class TokenManager implements ITokenManager {
+  private alg = 'RS256' as const
+  private grantType = 'client_credentials' as const
+  private clientAssertionType =
+    'urn:ietf:params:oauth:client-assertion-type:jwt-bearer' as const
+  private clientId: string
+  private clientKey: string
+  private keyId: string
+  private scope: string
+  private authServer: string
+  private httpClient: HTTPClient
+  private maxRetries: number
+  private clockSkewInSeconds = 0
+
+  private accessToken?: AccessToken
+  private tokenEmitter = new Emitter<{
+    access_token: [{ token: AccessToken } | { error: unknown }]
+  }>()
+  private retryCount: number
+  private pollerTimer?: ReturnType<typeof setTimeout>
+
+  constructor(props: OAuthSettings) {
+    this.keyId = props.keyId
+    this.clientId = props.clientId
+    this.clientKey = props.clientKey
+    this.authServer = props.authServer ?? 'https://oauth2.segment.io'
+    this.scope = props.scope ?? 'tracking_api:write'
+    this.httpClient = props.httpClient ?? new FetchHTTPClient()
```

Comment:
This should never be undefined?

Reply:
I don't think it's unreasonable for TokenManager to have defaults that publisher doesn't currently use.  I feel there's less harm in leaving it this way.

Reply:
@MichaelGHSeg I think it's a bit of a smell because it's not clear looking at the codebase _where_ the default is set. It's a bit of a sign when you are implementing (or in this case, handling) part of an interface that you're not using, that you need a different interface.

Reply:
<img width="770" alt="image" src="https://github.com/segmentio/analytics-next/assets/5115498/613a67b6-daf9-46ff-b1b1-f0ac18251853">